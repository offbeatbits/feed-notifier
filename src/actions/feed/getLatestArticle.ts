export const getLatestArticle = () => {
  console.log('getLatestArticle'); // eslint-disable-line
};
