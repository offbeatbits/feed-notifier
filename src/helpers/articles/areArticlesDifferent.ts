import { logger } from "../../lib/logger";
import { ArticleObject } from "../../types";

const areStringsDifferent = (
  someString: string | null | undefined,
  otherString: string | null | undefined
) => {
  if (someString && otherString) {
    return someString.localeCompare(otherString) !== 0;
  }
  return false;
};

const areDatesDifferent = (
  oldDate: string | null | undefined,
  newDate: string | null | undefined
) => {
  if (oldDate && newDate) {
    return (
      new Date(oldDate!).getTime() !== new Date(newDate).getTime() ||
      Boolean(oldDate)
    );
  }
  return false;
};

interface ArticlesToCompare {
  oldArticle: ArticleObject | null | undefined;
  newArticle: ArticleObject | null | undefined;
  skip?: boolean;
}

export const areArticlesDifferent = ({
  oldArticle,
  newArticle,
  skip,
}: ArticlesToCompare) => {
  if (skip) {
    logger.verbose("Skipping article comparison due to 'skipCache' flag...");
    return true;
  }

  logger.verbose("Comparing articles...");

  const titleIsDifferent = areStringsDifferent(
    oldArticle?.title,
    newArticle?.title
  );

  logger.verbose(`Article title is different: ${titleIsDifferent}`);

  const urlIsDifferent = areStringsDifferent(oldArticle?.url, newArticle?.url);

  logger.verbose(`Article URL is different: ${urlIsDifferent}`);

  const articleDateIsLater = areDatesDifferent(
    oldArticle?.date,
    newArticle?.date
  );

  logger.verbose(`Article date is later: ${articleDateIsLater}`);

  const feedUpdateDateIsLater = areDatesDifferent(
    oldArticle?.feedUpdateDate,
    newArticle?.feedUpdateDate
  );

  logger.verbose(`Feed build date is later: ${feedUpdateDateIsLater}`);

  const result =
    titleIsDifferent &&
    urlIsDifferent &&
    articleDateIsLater &&
    feedUpdateDateIsLater;

  logger.verbose(`Articles are different: ${result}`);

  return result;
};
