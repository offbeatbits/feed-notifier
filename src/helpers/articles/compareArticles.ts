import { logger } from '../../lib/logger';
import { ArticleObject } from '../../types';

const compareStrings = (
  someString: string | null | undefined,
  otherString: string | null | undefined,
) => {
  if (someString && otherString) {
    return someString.localeCompare(otherString) === 0;
  }
  return false;
};

const compareDates = (
  oldDate: string | null | undefined,
  newDate: string | null | undefined,
) => {
  if (oldDate && newDate) {
    return new Date(oldDate!).getTime() < new Date(newDate).getTime()
      || Boolean(oldDate);
  }
  return false;
};

interface ArticlesToCompare {
  oldArticle: ArticleObject | null | undefined;
  newArticle: ArticleObject | null | undefined;
  skip?: boolean;
}

export const compareArticles = ({
  oldArticle,
  newArticle,
  skip,
}: ArticlesToCompare) => {
  if (skip) {
    logger.verbose("Skipping article comparison due to 'skipCache' flag...");
    return true;
  }

  logger.verbose('Comparing articles...');

  const titleIsDifferent = compareStrings(
    oldArticle?.title,
    newArticle?.title,
  );

  logger.verbose(`Article title is different: ${titleIsDifferent}`);

  const urlIsDifferent = compareStrings(
    oldArticle?.url,
    newArticle?.url,
  );

  logger.verbose(`Article URL is different: ${urlIsDifferent}`);

  const articleDateIsLater = compareDates(
    oldArticle?.date,
    newArticle?.date,
  );

  logger.verbose(`Article date is later: ${articleDateIsLater}`);

  const feedUpdateDateIsLater = compareDates(
    oldArticle?.feedUpdateDate,
    newArticle?.feedUpdateDate,
  );

  logger.verbose(`Feed build date is later: ${feedUpdateDateIsLater}`);

  const result = titleIsDifferent
    && urlIsDifferent
    && articleDateIsLater
    && feedUpdateDateIsLater;

  logger.verbose(`Articles are different: ${result}`);

  return result;
};
