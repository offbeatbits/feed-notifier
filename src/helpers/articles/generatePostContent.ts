import { logger } from "../../lib/logger";
import { ArticleObject } from "../../types.d";

export const generatePostContent = (article: ArticleObject) => {
  const content = `${article.title} ${article.url}`;
  logger.verbose(`Status content: ${content}`);
  return content;
};
