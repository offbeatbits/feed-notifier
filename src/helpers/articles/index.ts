export * from './compareArticles';
export * from './generatePostContent';
export * from './getCachedArticle';
export * from './getLatestArticle';
export * from './updateArticleCache';
