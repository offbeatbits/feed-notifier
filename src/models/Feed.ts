import Parser from 'rss-parser';

export class Feed {
  url: string;

  private content: any;

  constructor(url: string) {
    this.url = url;
  }

  async fetch() {
    this.content = await new Parser().parseURL(this.url);
    return this;
  }

  getLatestPostDate() {
    return this.content.items[0].pubDate;
  }

  getLatestPostTitle() {
    return this.content.items[0].title;
  }

  getLatestPostUrl() {
    return this.content.items[0].link;
  }

  getLatestPostId() {
    return this.content.items[0].id;
  }
}
