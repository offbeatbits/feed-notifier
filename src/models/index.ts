export * from './Channel';
export * from './Feed';
export * from './OAuth2Service';
