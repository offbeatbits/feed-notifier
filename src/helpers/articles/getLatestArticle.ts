import { logger } from '../../lib/logger';
import { fetchFeed } from '../feed';
import { ArticleObject } from '../../types';

export const getLatestArticle = async (feedUrl: string) => {
  logger.verbose('Getting latest article...');
  try {
    const feed = await fetchFeed(feedUrl);

    const article = {
      feedUpdateDate: feed.getFeedUpdateDate() || feed.getLatestPostDate(),
      date: feed.getLatestPostDate(),
      title: feed.getLatestPostTitle(),
      url: feed.getLatestPostUrl(),
    } as ArticleObject;

    logger.verbose(`Feed update date: ${article.feedUpdateDate}`);
    logger.verbose(`Latest article date: ${article.date}`);
    logger.verbose(`Latest article title: ${article.title}`);
    logger.verbose(`Latest article URL: ${article.url}`);

    return article;
  } catch {
    logger.verbose('Something went wrong while fetching feed, returning null!');
    return null;
  }
};
