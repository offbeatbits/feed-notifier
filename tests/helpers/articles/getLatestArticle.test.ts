import { Feed } from '../../../src/models/Feed';
import { getLatestArticle } from '../../../src/helpers/articles/getLatestArticle';

jest.mock('rss-parser', () => jest.fn().mockImplementation(() => ({
  parseURL: () => Promise.resolve({
    content: {
      lastBuildDate: 'gla_mocked_last_build_date',
    },
    lastBuildDate: 'gla_mocked_last_build_date',
    items: [
      {
        pubDate: 'gla_mocked_last_item_pub_date',
        title: 'gla_mocked_last_item_title',
        link: 'gla_mocked_last_item_link',
      },
    ],
  }),
})));

describe('getLatestArticle', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should get latest article', async () => {
    expect.assertions(1);
    await expect(getLatestArticle('sample_feed_url')).toMatchSnapshot();
  });

  it('should return null if feed returns no article', async () => {
    expect.assertions(1);
    (Feed.prototype as any).fetch = () => Promise.resolve(null);
    await expect(getLatestArticle('sample_feed_url')).toBeNull();
  });
});
