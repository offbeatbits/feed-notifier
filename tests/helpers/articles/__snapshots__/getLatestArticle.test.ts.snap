// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getLatestArticle should get latest article 1`] = `
Object {
  "date": "gla_mocked_last_item_pub_date",
  "feedUpdateDate": "gla_mocked_last_build_date",
  "title": "gla_mocked_last_item_title",
  "url": "gla_mocked_last_item_link",
}
`;
